Feature/button animation

AliNisarAhmed opened the pull request:
## Description

Adds wave like Button animation upon click

## Action Items

- [x] I have read the [CONTRIBUTING](../CONTRIBUTING.md) doc
- [x] This PR is as small as it can be, and only implements on thing (not many)
- [x] I have manually tested my feature
- [ ] I have written a "good" amount of tests (including visual and unit tests)

AliNisarAhmed commented:
This PR is still Work In Progress... I created it to get some feedback.

I used elm-css to create the animation... when the button is `active` the `::before` pseudo element is doing the wave like animation.

Some questions on which help would be appreciated.

1. On lines 161-163, im manually defining the `transform` property to animate. In `elm-css`'s `Css.Animations` module, there is a `transform` function for doing that, but its type is confusing me. It has the type 
  ```
  transform : List { compatible | value : String, transform : Compatible } -> Property

  ```
what value do I provide it to animate it? What is Compatible type, and how to create a record with those fields is unclear.

2. In `elm-css`, how do I apply `animation-timing-function` and `animation-fill-mode`? I tried adding a `transition` to achieve the peculiar cubic bezier timing function of antd, but it is not working.

3. Ideally, I would like to start the animation on the event `onMouseUp`. (That is how antd is doing it, by making a custom HTML attribute value to `true` or `false`). Any ideas how I can achieve the same in Elm.

Any help would be greatly appreciated.

Orasund commented:
>1. On lines 161-163, im manually defining the transform property to animate. In elm-css's Css.Animations module, there is a transform function for doing that, but its type is confusing me. It has the type
>````transform : List { compatible | value : String, transform : Compatible } -> Property````
>
> what value do I provide it to animate it? What is Compatible type, and how to create a record with those fields is unclear.

This is just a technical detail, it just says "anything that is a transition can be used." In Particular, anything that is of type `Transition`. Here is also a example in the docs how that should look like:

```
transition
            [ Css.Transitions.backgroundColor 1000
            , Css.Transitions.transform3 500 0 easeInOut
            ]
```

>2. In elm-css, how do I apply animation-timing-function and animation-fill-mode? I tried adding a transition to achieve the peculiar cubic bezier timing function of antd, but it is not working.

Im guessing that is currently not supported and you will have to define them yourself using `Css.property`

>3. Ideally, I would like to start the animation on the event onMouseUp. (That is how antd is doing it, by making a custom HTML attribute value to true or false). Any ideas how I can achieve the same in Elm.

I believe you will need to give the button an id, and then once a OnMouseUp Event is triggered the code for the animation is added to the button (similarly OnMouseOut will remove the code for the animation again).

supermacro commented:
I've added previews to all PRs! You can click on the `deploy/netlify` CI check to view a live version of the showcase application.

AliNisarAhmed commented:
@supermacro I don't think this is possible without adding some JS. My plan is to add a class/attribute on mouse up, and set a timeout to remove it after the animation is done... if I am able to set a class/attribute with JS, I may be able to fully replicate antd's way of doing it without even using `::before` pseudo element.

Any ideas how I can add JS?

supermacro commented:
I thought that you could use `Html.property` or `Html.attribute` to add [events](https://html.com/javascript/events/), but I tried it out and it does not work. 

An alternative might be to have a external JS script that adds additional functionality. There might be a way to programattiaclly add this JS script to the DOM (similar to how [Elm Bootstrap](https://package.elm-lang.org/packages/EdutainmentLIVE/elm-bootstrap/latest/) loads its styles programmatically from a CDN). Otherwise, we would need to add documentation to inform users that they must add a JS script to their app in order to attain extra animations - and I'm totally ok with this approach.

Orasund commented:
Im sceptical  that will work without any ugly workaround.

Typically trying to access the Dom outside Elm leads to race conditions, because Elm loves to do a lot of unexpected things under the hood. That's why Html.Lazy exists. As you might know, Html.Lazy requires an Id for any Dom element that should be accessable from outside Elm.

I think there is nothing wrong with a pure elm solution. The simplest one would be to store `mouseOver: Maybe String` in the model.

supermacro commented:
> I think there is nothing wrong with a pure elm solution. The simplest one would be to give the button an id and to store mouseOver: Maybe String in the model.

I am ok with this so long as it's opt-in. I.e. if I don't care about the animation as a user, and I don't want to add more code to my model, then I shouldn't be forced to add `mouseOver: Maybe String` to my model in order to use `Button`

AliNisarAhmed commented:
@Orasund if you have any ideas on how to do this in pure elm, please feel free to take over and implement it (I will be more than happy to learn). I have not been able to make any progress coz (a) I do not have time this week (b) I am out of ideas anyways...

AliNisarAhmed commented:
Someone on elm slack shared this example https://ellie-app.com/9qtRjwDj6SVa1 if it helps...

Orasund commented:
This example uses elm-animator.

I can give an example for pure CSD+Elm on the weekend (i only got my phone during the week).

AliNisarAhmed commented:
Hey, some good news, I got an initial pure elm solution working ([see it here](https://github.com/AliNisarAhmed/elm-animated-button)). It starts animation on mouse up...

Next step will be to create and expose functions from the button module for users to incorporate in their own `update` function, and also maybe expose a ButtonMsg opaque type...

Orasund commented:
Yeah, that exactly what I meant.

You can switch `OnMouseUp` with `OnMouseOver` to get the result you are looking for.

supermacro commented:
Just to make sure we're on the same page, let me re-post a recent comment of mine:

> I am ok with this so long as it's opt-in. I.e. [As an Elm Antd user] if I don't care about the animation, and I don't want to add more code to my model, then I shouldn't be forced to add `xyz` to my model in order to use Button

AliNisarAhmed commented:
Update: I now have a pure elm solution (in my separate [branch](https://github.com/AliNisarAhmed/elm-animated-button), see the deployed version on netlify [here](https://compassionate-swirles-0d66af.netlify.app/)) which would work in the following way

1. If the users do not want animation, they will continue to use the `button` function from `elm-antd`
2. However, if they do want an animated button, they will have available to them a `withAnimation` function, which takes in a msg, model and a normal button and returns an animated button.
  a. However, to have animation, for each animated button, the users will also have to handle the following things: 
    - provide a Model data constructor with `ButtonModel` type argument exported from the library.
    - provide Msg data constructor, with `ButtonMsg` type argument exported from the library.
    - in their `update` function, handle updates for each animated button using the exported `buttonUpdater` function (passing in the respective model and mapping the commands)
 
@supermacro Let me know what you think of this API exposed to the users. I think this is the minimum that the users must handle in order to have a pure elm onMouseUp animation, because everything must go through the central update function (which is both a strength and kinda weakness of Elm Architecture, I so wish there was a local update function and local state possible).

We can avoid all this with use of JS, but the current project structure needs to be changed to accomodate JS code, which I think @supermacro is in the best position to decide.

Please let me know what you think, and I will start incorporating this API in `elm-antd`

Orasund commented:
Is a button that was created by

```
button
|> withAnimation AnimateMyButton {animate : false}
```

a valid button or is that an impossible state? I would expect it to be the latter.

Maybe a better type signature would be

`withAnimation: Maybe msg -> ButtonModel msg -> ButtonModel msg`

Other than that it looks great.

AliNisarAhmed commented:
@Orasund I do not understand what you mean by "valid button or an impossible state". Can you please elaborate more what you mean?

An animated button can only be created with the use of `withAnimation` function, so maybe in a way that is kind of an impossible state for our button, in the sense that user cannot reach `AnimatedButton` without the help of our function, which is exactly as intended. The animation is a conscious opt-in by the user of the library.

As an aside, thinking about this more, I think all other functions exposed by `elm-antd` that operate on Normal Button would mostly remain the same, since `AnimatedButton` also takes the `Options msg` type argument, so all the operations on that options record are valid for both buttons.

Orasund commented:
I was asking if there exists a scenario where I would want to pass a msg but not animate the button.

To answer my own question: I now believe the answer is Yes.

`withAnimation MyMsg false` means "this button may be animated, but right now it's not and by hovering over it, it will trigger MyMsg".

It took me a while to get there, but now I believe this is actually a very clean solution.

AliNisarAhmed commented:
@Orasund One thing I want to clarify is that the Ant Design animation I'm trying to replicate here happens on `onMouseUp` event, not on `:hover` and not on `:active`. 

If you zoom in on the `antd` button [here](https://ant.design/components/button/), when the button is `:active`, the button changes color to darker share of primary color (which is already there in `elm-antd`). What this PR is implementing is the wave animation that happens when the user lifts the mouse up...

supermacro commented:
Hey @AliNisarAhmed this is awesome!

Just finished reading through the code and it makes sense to me and I think it's a good design. 

However, one of my guiding principles for this library is ease of use. I do not think the benefit of having native-elm animations outweighs the cost of the boilerplate necessary to have them in a user's project. 

Therefore I still think it's better to create a dedicated JS file for "extra" functionality such as animations that a user can opt into. 

So what I think would make sense is to have a new folder called `elm-andt-extras` that contains a JS / webpack project that is then bundled into a single minified js file. That file would then be accessible via [JSDeliver](https://www.jsdelivr.com/) or some other npm cdn.

AliNisarAhmed commented:
@supermacro I totally agree with you. Elm architecture, while it makes majority of the things super easy, makes small things (like this animation) cumbersome by diverting every single thing through one update function... 

I agree, better DX will be to do this in JS... I will create the folder and try to set up a webpack project... I will see how far I go in setting that up (have never done this), and if stuck I will ask for help here...

supermacro commented:
@AliNisarAhmed for v1 / simplicity, you can do the following:

- Do not spend time on webpack stuff
- Create a single `index.js` file inside of `elm-antd-extras`
- Figure out how to get the showcase app's `index.html` (located [here](https://github.com/supermacro/elm-antd/blob/master/showcase/src/index.html)) to load this `index.js` file.

AliNisarAhmed commented:
`npm run dev` in showcase is giving an error "FILE_SERVER_URL" is not recognized as an internal or external command. I am on Windows.

supermacro commented:
Hey @AliNisarAhmed, I hadn't considered cross-platform issues with my scripts.

The `dev` script creates an environment variable for the running process called `FILE_SERVER_URL`. 

There's a few solutions to this problem. The best one I think is to remove the declaration of `FILE_SERVER_URL` and instead do the following in whatever JS file uses it:

`process.env.FILE_SERVER_URL || 'http://localhost:8080'`

AliNisarAhmed commented:
@supermacro check out the PR now, I have added a JS file, which selects all buttons with a class, and then adds a class to animate, before removing it with a setTimeout.

I have added an `index.js` file in `elm-antd-extras` folder, which is then included in the project using webpack (dont know if this is the right way to do it, but it works).

supermacro commented on elm-antd-extras/index.js:
```diff
@@ -0,0 +1,10 @@
+const buttons = document.querySelectorAll('.animated-btn');
```
Use the following convention for classes:

`elm-antd__<snake_case_identifier>`. 

So in this case it would be, `elm-antd__animated_button`

supermacro commented on showcase/webpack.config.js:
```diff
@@ -25,10 +25,11 @@ const commonConfig = {
     app: './src/index.js',
-    sw: './src/sw.js'
+    sw: './src/sw.js',
+    extra: '../elm-antd-extras'
```
Yup, this works fine for the showcase! Although I am wondering if there's a way to specify that the `extra` bundle be injected into the `head` of the html instead of the bottom of the `body` ... but that's not a big deal right now.

supermacro commented on src/Ant/Button.elm:
```diff
@@ -153,6 +155,40 @@ toHtml (Button options label) =
+                ]
+
+        animationStyle =
+            CG.withClass "animated-before" <| [ position relative, animatedBefore (hex primaryColorStrong) ]
```
See comment regarding class naming conventions.

supermacro commented on src/Ant/Button.elm:
```diff
@@ -296,10 +338,10 @@ toHtml (Button options label) =
                 Just msg ->
-                    [ A.disabled options.disabled, Events.onClick msg, css <| cursorHoverStyles :: combinedButtonStyles ]
+                    [ A.class "animated-btn", A.disabled options.disabled, Events.onClick msg, css <| cursorHoverStyles :: combinedButtonStyles ]
 
                 Nothing ->
-                    [ A.disabled options.disabled, css <| cursorHoverStyles :: combinedButtonStyles ]
+                    [ A.class "animated-btn", A.disabled options.disabled, css <| cursorHoverStyles :: combinedButtonStyles ]
```
Please update class names

supermacro commented:
Seeing some inconsistencies in behaviour between the button in https://ant.design/components/button/ and the button implemented here. 

FYI: All PRs have a dedicated showcase deployment on netlify (you can find it by clicking on "Details" within the `deploy/netlify` ci check). The one for this PR is https://deploy-preview-31--elm-antd.netlify.app/components/button

Here's a screen recording of the behaviour I'm seeing:  https://www.loom.com/share/328154eef6274faaab18f2e6e4e36911

When I click the first time, the wave animation appears. But when I click a second time, the animation only sometimes appears and it looks like I have to move out of the mouse and back into it to see the animation consistently.

supermacro commented:
I wonder if this behaviour is due to the Elm runtime behaving in a manner which we don't fully understand. There's a comment that @Orasund made recently about that. 

https://github.com/supermacro/elm-antd/pull/31#issuecomment-657952768

AliNisarAhmed commented:
@supermacro yeah I noticed that too. I think it is because of the setTimeout. If they click the button again I should clear the timeout and remove the animation class as well... 

Working on this plus class naming conventions as well...

Orasund commented:
> I wonder if this behaviour is due to the Elm runtime behaving in a manner which we don't fully understand. There's a comment that @Orasund made recently about that.
> 
> [#31 (comment)](https://github.com/supermacro/elm-antd/pull/31#issuecomment-657952768)

It seems that Elm is not the problem here. (I used DevTools for Chrome to check that Elm isn't removing the Button when updating)

I also checked the case where Elm is actually the problem. Here is what i did (i sadly didn't keep the ellie):

AliNisarAhmed commented:
@Orasund @supermacro  I pushed a fix, seems pretty close to antd's now... please check.

supermacro commented:
Oh wow that's amazing 😍

supermacro commented on src/Ant/Button.elm:
```diff
@@ -296,10 +338,10 @@ toHtml (Button options label) =
         attributes =
             case options.onClick of
                 Just msg ->
-                    [ A.disabled options.disabled, Events.onClick msg, css <| cursorHoverStyles :: combinedButtonStyles ]
+                    [ A.class "elm-antd__animated_btn", A.disabled options.disabled, Events.onClick msg, css <| cursorHoverStyles :: combinedButtonStyles ]
 
                 Nothing ->
-                    [ A.disabled options.disabled, css <| cursorHoverStyles :: combinedButtonStyles ]
+                    [ A.class "elm-antd__animated_btn", A.disabled options.disabled, css <| cursorHoverStyles :: combinedButtonStyles ]
```
Could u please refactor this section? 

```elm
baseAttributes =
    [ A.class "elm-antd__animated_btn"
    , A.disabled options.disabled
    , Events.onClick msg
    , css <| cursorHoverStyles :: combinedButtonStyles 
    ]

attributes =
    case options.onClick of
        Just msg ->
            Events.onClick msg :: baseAttributes
        
        Nothing ->
            baseAttributes 
```

Don't forget to run `scripts/format-elm-files.sh` :)

AliNisarAhmed replied:
done. 

I am using elm-format so don't think I need to run format script :)

AliNisarAhmed commented:
I resolved the comments, please let me know what else is needed to merge this PR.
